src/controllers/routers/search.js

```diff
@@ -0,0 +1,134 @@
+/* eslint-disable camelcase */
+const { raw } = require('objection');
+const { safeResource } = require('../../helpers/cleandata/removeid');
+const { Beneficiary, Referee, Case } = require('../../models');
+const {
+  errors: { BadRequest },
+} = require('../../utils');
+
+const search = async (req, res) => {
+  // Extract all query params
+  const {
+    q,
+    type,
+    with_paging,
+    page,
+    per_page,
+    fields,
+    include_entities,
+  } = req.query;
+  // Choose database based on `type` query param
+  let model;
+  let searchFields;
+  if (type === 'beneficiary') {
+    model = Beneficiary;
+    searchFields = ['beneficiary."name"', 'beneficiary."email"'];
+  } else if (type === 'case') {
+    model = Case;
+    searchFields = ['"caseNumber"'];
+  } else if (type === 'referee') {
+    model = Referee;
+    searchFields = [
+      'referee."name"',
+      'referee."email"',
+      'referee."organisation"',
+    ];
+  } else {
+    return res.json(new BadRequest(`Type is missing.`));
+  }
+
+  // TODO : create custom join login based on entities.
+  const searchBody = searchFields.join(`||`);
+  const entities = include_entities ? include_entities.split('&') : [];
+  const sqlQuery = `similarity (':searchBody:', ${searchBody}) > 0.08`;
+  let fetchWith;
+  let order;
+
+  if (type === 'beneficiary' && entities.includes('cases')) {
+    fetchWith = `cases(caseNumber)`;
+    order = `similarity(beneficiary."name", '${q}') DESC`;
+  }
+  if (type === 'beneficiary' && entities.includes('referees')) {
+    fetchWith = '[referees]';
+    order = `similarity(beneficiary."name", '${q}') DESC`;
+  }
+
+  if (type === 'referee' && entities.includes('cases')) {
+    fetchWith = `cases(caseNumber)`;
+    order = `similarity(referee."name", '${q}') DESC`;
+  }
+
+  if (type === 'referee' && entities.includes('beneficiary')) {
+    fetchWith = '[beneficiary]';
+    order = `similarity(referee."name", '${q}') DESC`;
```

Comment:
Can you find a way to locate all the type-specific logic in one place, as in line 23